packages/jellyfish-api-core/__tests__/category/loan/takeLoan.test.ts

```diff
@@ -3,279 +3,636 @@ import { GenesisKeys } from '@defichain/testcontainers'
+  })
 
-    await tGroup.get(0).rpc.loan.depositToVault({
-      vaultId: vaultId1, from: collateralAddress, amount: '12000@DFI'
+  it('should not takeLoan on mintable:false token', async () => {
+    await bob.container.call('updateloantoken', ['TSLA', { mintable: false }])
+    await bob.generate(1)
+
+    const promise = bob.rpc.loan.takeLoan({
+      vaultId: bobVaultId,
+      amounts: '30@TSLA'
     })
-    await tGroup.get(0).generate(1)
+    await expect(promise).rejects.toThrow(RpcApiError)
+    await expect(promise).rejects.toThrow('Loan cannot be taken on token with id (2) as "mintable" is currently false')
+
+    await bob.container.call('updateloantoken', ['TSLA', { mintable: true }])
+    await bob.generate(1)
     await tGroup.waitForSync()
-  }
+  })
 
-  describe('takeLoan', () => {
-    beforeAll(async () => {
-      await tGroup.start()
-      await tGroup.get(0).container.waitForWalletCoinbaseMaturity()
-      await setup()
+  it('should not takeLoan on inLiquidation vault', async () => {
+    await bob.rpc.loan.takeLoan({
+      vaultId: liqVaultId,
+      amounts: '2000@UBER'
     })
+    await bob.generate(1)
+    await tGroup.waitForSync()
 
-    afterAll(async () => {
-      await tGroup.stop()
-    })
+    await alice.rpc.oracle.setOracleData(
+      oracleId,
+      timestamp,
+      { prices: [{ tokenAmount: '6@UBER', currency: 'USD' }] })
+    await alice.generate(12)
+    await tGroup.waitForSync()
 
-    it('should not takeLoan on nonexistent vault', async () => {
-      const promise = tGroup.get(0).rpc.loan.takeLoan({
-        vaultId: '0'.repeat(64),
-        amounts: '30@TSLA'
-      })
-      await expect(promise).rejects.toThrow(RpcApiError)
-      await expect(promise).rejects.toThrow(`Vault <${'0'.repeat(64)}> not found`)
-    })
+    const liqVault = await bob.rpc.loan.getVault(liqVaultId) as VaultLiquidation
+    expect(liqVault.state).toStrictEqual('inLiquidation')
 
-    it('should not takeLoan on nonexistent loan token', async () => {
-      const promise = tGroup.get(0).rpc.loan.takeLoan({
-        vaultId: vaultId,
-        amounts: '1@BTC'
-      })
-      await expect(promise).rejects.toThrow(RpcApiError)
-      await expect(promise).rejects.toThrow('Loan token with id (1) does not exist!')
+    const promise = bob.rpc.loan.takeLoan({
+      vaultId: liqVaultId,
+      amounts: '1@UBER'
     })
+    await expect(promise).rejects.toThrow(RpcApiError)
+    await expect(promise).rejects.toThrow('Cannot take loan on vault under liquidation')
+  })
 
-    it('should not takeLoan on invalid token', async () => {
-      const promise = tGroup.get(0).rpc.loan.takeLoan({
-        vaultId: vaultId,
-        amounts: '1@INVALID'
-      })
-      await expect(promise).rejects.toThrow(RpcApiError)
-      await expect(promise).rejects.toThrow('Invalid Defi token: INVALID')
+  it('should not takeLoan on mayLiquidate vault', async () => {
```

Comment:
This test is perhaps less meaningful, `takeLoan` will still go through as long as resulted total amount within ratio.

Suggest test to be added
`should not takeLoan if total loan exceeded min collateral ratio` with taking multiple loan token but each within ratio.

Reply:
the mayLiquidate test can be used to assert the error message
and agree on testing the `liquidated by take multi loans`, it has extra logic... added